Scaffolding and implementation for new `PublicKey.Signature` API

jhenahan opened the pull request:
Contains the changes from #153, along with substantial additions. Happy to break this up into multiple PRs after some review, it's just harder to motivate without the whole picture. If the style is agreeable (and CI passes happily enough), I'd like to push in a similar direction for the other modules as followup work.

Highlights:
- Adds an internal `Codensity` called `Scoped` to clean up nested CPS code
- Adds utilities for serialization and deserialization of key material, along with instance utilities and error modeling
- Adds constant-time comparison `Eq` for safety against timing attacks on secret material
- Migrates all uses of `memcpy` and `memcmp` to `Foreign.copyBytes` and `Data.ByteString.Internal.memcmp`, respectively
- Hides all the internals of `PublicKey.Signature`
- Expanded test suite for `PublicKey.Signature`
- Lots of docs

Kleidukos commented:
@jhenahan This looks promising. Now we have to remove some stuff, but you got the spirit. :)

jhenahan commented on sel/src/Sel/ByteString/Codec.hs:
```diff
@@ -0,0 +1,161 @@
+-- serialization, transmission, display, etc.
+--
+-- @since 0.0.3.0
+class HexEncode k where
```
For sure. The syntactic convenience didn't feel particularly worth the lawless classes (though the deriving stuff that came with it felt cute), so I'll either inline these or come up with a light value-level equivalent.

Kleidukos replied:
What we had before suited me perfectly in terms of architecture. Don't make your life harder with type classes, good old functions still fit the bill. :)

jhenahan commented on sel/src/Sel/Internal/FFI.hs:
Yeah, I was torn about this. I opted to add the `memcpy` because `memcmp` was already in the previous code, and because all the other existing names I could find were using `ccall` rather than `capi`. Happy to just use the others, if that would be preferred. Would it be worth shifting all the `CUChar` signatures to `Word8` while we're at it? I had a poke around at some other code that was all in on `ByteArray#` and the like, as well, but that felt like it might be a bit outside the "teaching library" mandate.

Kleidukos replied:
> Happy to just use the others, if that would be preferred 

Yes I think so.

> Would it be worth shifting all the CUChar signatures to Word8 while we're at it?

No let's keep `CUChar`. You can wrap / unwrap the newtype as needed. :)

jhenahan commented on sel/sel.cabal:
Mostly a helper while I was churning modules, but if you like the style then I'm happy to automate it.

Kleidukos replied:
Let's do the migration in a separate PR. :)

jhenahan commented:
Dropped codec classes in favor of HOFs (though I can also strip that indirection if you'd rather not deal with it), dropped cabal formatting, migrated to `Foreign.copyBytes` and `Data.ByteString.Internal.memcpy`. Also dropped the negative instances since getting them to work across the supported range was not gonna happen. Will save such ideas for when `Unsatisfiable` is more reliably at hand.

jhenahan commented:
API diff against `main`

https://gist.github.com/jhenahan/c4938656e421bd7bc3e115aa74f8a1b2

Kleidukos commented on sel/src/Sel/Internal/Instances.hs:
This module should be merged back into `Sel.Internal`

Kleidukos commented on sel/src/Sel/ByteString/Codec/KeyPointer.hs:
I'm not convinced this needed at the moment.

Kleidukos commented on sel/src/Sel/ByteString/Codec.hs:
Let's not generalise the concept of Codec in this PR.
